Simple, portable LinuxPWM using sysfs API

RobertPHeller opened the pull request:
This is a simple LinuxPWM.hxx using the sysfs API.  It has been lightly tested (I wired up a breadboard with two LEDs connected to my 'Pi and used a simple OpenMRN node program to set different PWM duty cycles and the LEDs lit at the corresponding brightnesses).  It is just a simple derived class from PWM.hxx -- no fun and games with templates or other exciting things.

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,140 @@
+class LinuxPWM : public PWM
+{
+public:
+    LinuxPWM(int chip, int channel) : channel_(channel), chip_(chip)
```
it might make sense to swap the chip and channel variable assignments (and declarations in the class) to match the constructor arg order.  I'd also suggest making the variable declarations as const as they do not change in the lifetime of this class instance.

RobertPHeller replied:
Done. Updated version tested, commited, and pushed.

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,140 @@
+    {
+        char periodname[60];
+        uint32_t counts;
+        snprintf(periodname,sizeof(periodname),"/sys/class/pwm/pwmchip%d/pwm%d/period",chip_,channel_);
```
would it make sense to create a const std::string of "/sys/class/pwm/pwmchip%d/pwm%d" in the constructor that you then use here instead of snprintf in every method since the base value doesn't appear to be changing from each method, just the segment after this.

RobertPHeller replied:
I did this, but used a plain char[] -- using std::string seemed like overkill.  Also tested, commited, and pushed.

atanisoft replied:
I agree it may be overkill

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,174 @@
+private:
+    const uint32_t chip_;
+    const uint32_t channel_;
+    char pwmdir[40];
```
this should be pwmdir_, it may also be possible to make this const but I'm not sure.

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,174 @@
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_WRONLY);
```
is there a reason you opted for mixing fopen and open? it might make more sense to use just one api.

RobertPHeller replied:
Only that when I am writing a pure constant (eg for the enable/disable functions), I don't need the overhead of fprintf().  But where I need to convert a number to a string, I need fprintf() (or snprintf() if I used open()...).  Why? Does it really matter?  Remember, this is on machines running a fullfledged OS (Linux), not on MCUs running a minimualist OS (FreeRTOS) out of Flash memorty.

atanisoft replied:
Fair enough and makes some sense. There may even be other options available in OpenMRN that I haven't explored yet as well.

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,174 @@
+        char exportname[50];
+        strcpy(exportname,pwmdir);
+        strcat(exportname,"/period");
+        FILE *fp = fopen(exportname,"r");
```
instead of fopen you can use
```
if (access(exportname, F_OK) != -1)
{
 // it exists
}
```

atanisoft replied:
you can also use W_OK or R_OK to test for write or read access.

RobertPHeller replied:
Ah, I knew there was a simple test, but I guess my C lib memory is a bit rusty...

balazsracz commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,173 @@
+    {
+        char periodname[60];
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
```
this code and around is duplicated a lot of times.
I would recommend two helper functions:

```
private:
uint32_t get_sysfs_file_value(const char* basename) {
...
}
void set_sysfs_file_value(const char* basename, uint32_t value) {
...
}
```
then most of the copies of this code could be removed.

RobertPHeller commented:
At Sun, 03 Mar 2019 07:47:04 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:

> 
> 
> balazsracz commented on this pull request.
> 
> 
> 
> > +        }
> +        char *p = strrchr(exportname,'/');
> +        *p = '\0';
> +        p = strrchr(exportname,'/');
> +        p++;
> +        strcpy(p,"export");
> +        FILE *fp = fopen(exportname,"w");
> +        fprintf(fp,"%d\n",channel_);
> +        fclose(fp);
> +        set_period(1);
> +    }       
> +    void set_period(uint32_t counts) override
> +    {
> +        char periodname[60];
> +        strcpy(periodname,pwmdir);
> +        strcat(periodname,"/period");
> 
> this code and around is duplicated a lot of times.
> I would recommend two helper functions:

Done.  Tested and pushed.


> 
> ```
> private:
> uint32_t get_sysfs_file_value(const char* basename) {
> ...
> }
> void set_sysfs_file_value(const char* basename, uint32_t value) {
> ...
> }
> ```
> then most of the copies of this code could be removed.
> 

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services

atanisoft commented on src/os/LinuxPWM.hxx:
```diff
@@ -0,0 +1,164 @@
+        return value;
+    }
+    void set_sysfs_file_value(const char* basename, uint32_t value) {
+        FILE *fp = fopen(basename,"w");
```
I believe @balazsracz intent was to move the strcpy/strcat into this (and method above) so that the callers would just be:
```
set_sysfs_file_value("enable", 1);
```
and internally it would be converted to pwmdir + "/" + "enable"

balazsracz replied:
see https://github.com/RobertPHeller/openmrn/pull/4
note you also need to merge openmrn master for it to work.

RobertPHeller commented:
At Mon, 04 Mar 2019 11:33:24 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:

> 
> 
> balazsracz commented on this pull request.
> 
> 
> 
> > +        strcat(enablename,"/enable");
> +        return get_sysfs_file_value(enablename) == 1;
> +    }
> +private:
> +    const uint32_t chip_;
> +    const uint32_t channel_;
> +    char pwmdir[40];
> +    uint32_t get_sysfs_file_value(const char* basename) {
> +        uint32_t value;
> +        FILE *fp = fopen(basename,"r");
> +        fscanf(fp,"%d",&value);
> +        fclose(fp);
> +        return value;
> +    }
> +    void set_sysfs_file_value(const char* basename, uint32_t value) {
> +        FILE *fp = fopen(basename,"w");
> 
> see https://github.com/RobertPHeller/openmrn/pull/4
> note you also need to merge openmrn master for it to work.
> 

What do I do to resolve this?  I don't know what to do.

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services

RobertPHeller commented:
OK, now things are totally broken.  It fails to compile with a link error:

PWMTest.o: In function `openlcb::GenMessage::GenMessage()':
/home/heller/openmrn/src/utils/FileUtils.hxx:69: multiple definition of `read_file_to_string(std::string const&)'
main.o:/home/heller/openmrn/src/utils/FileUtils.hxx:69: first defined here
PWMTest.o: In function `write_string_to_file(std::string const&, std::string const&)':
/home/heller/openmrn/src/utils/FileUtils.hxx:95: multiple definition of `write_string_to_file(std::string const&, std::string const&)'
main.o:/home/heller/openmrn/src/utils/FileUtils.hxx:95: first defined here
collect2: error: ld returned 1 exit status
/home/heller/openmrn/etc/prog.mk:187: recipe for target 'PWMTest_RPI' failed
make[2]: *** [PWMTest_RPI] Error 1
make[2]: Leaving directory '/home/heller/RRCircuits/PWMTest/targets/rpi.linux.armv7a'

Also, it appears my repo is 30 commits behind bakerstu/openmrn and I have no 
clue how to fix that.  GitHub will only let me make a pull request and is not 
giving me a way to get the updates.  I am about to completely trash my repo 
and make a fresh fork and start over.

I am toally confused by GitHub.


At Mon, 04 Mar 2019 11:33:24 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:

> 
> 
> balazsracz commented on this pull request.
> 
> 
> 
> > +        strcat(enablename,"/enable");
> +        return get_sysfs_file_value(enablename) == 1;
> +    }
> +private:
> +    const uint32_t chip_;
> +    const uint32_t channel_;
> +    char pwmdir[40];
> +    uint32_t get_sysfs_file_value(const char* basename) {
> +        uint32_t value;
> +        FILE *fp = fopen(basename,"r");
> +        fscanf(fp,"%d",&value);
> +        fclose(fp);
> +        return value;
> +    }
> +    void set_sysfs_file_value(const char* basename, uint32_t value) {
> +        FILE *fp = fopen(basename,"w");
> 
> see https://github.com/RobertPHeller/openmrn/pull/4
> note you also need to merge openmrn master for it to work.
> 

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services

atanisoft commented:
> Also, it appears my repo is 30 commits behind bakerstu/openmrn and I have no clue how to fix that. GitHub will only let me make a pull request and is not giving me a way to get the updates. I am about to completely trash my repo and make a fresh fork and start over.

You have a couple options:
1) https://github.com/KirstieJane/STEMMRoleModels/wiki/Syncing-your-fork-to-the-original-repository-via-the-browser
2) https://help.github.com/en/articles/syncing-a-fork

I use the second option normally as I use commandline client often.

RobertPHeller commented:
At Mon,  4 Mar 2019 16:15:05 -0500 (EST) Robert Heller <heller@deepsoft.com> wrote:

> 
> OK, now things are totally broken.  It fails to compile with a link error:
> 
> PWMTest.o: In function `openlcb::GenMessage::GenMessage()':
> /home/heller/openmrn/src/utils/FileUtils.hxx:69: multiple definition of `read_file_to_string(std::string const&)'
> main.o:/home/heller/openmrn/src/utils/FileUtils.hxx:69: first defined here
> PWMTest.o: In function `write_string_to_file(std::string const&, std::string const&)':
> /home/heller/openmrn/src/utils/FileUtils.hxx:95: multiple definition of `write_string_to_file(std::string const&, std::string const&)'
> main.o:/home/heller/openmrn/src/utils/FileUtils.hxx:95: first defined here
> collect2: error: ld returned 1 exit status
> /home/heller/openmrn/etc/prog.mk:187: recipe for target 'PWMTest_RPI' failed
> make[2]: *** [PWMTest_RPI] Error 1
> make[2]: Leaving directory '/home/heller/RRCircuits/PWMTest/targets/rpi.linux.armv7a'
> 
> Also, it appears my repo is 30 commits behind bakerstu/openmrn and I have no 
> clue how to fix that.  GitHub will only let me make a pull request and is not 
> giving me a way to get the updates.  I am about to completely trash my repo 
> and make a fresh fork and start over.
> 
> I am toally confused by GitHub.

OK, I *think* I have fixed things...

One other thing: the way the current sysfs interface work, you have to be root 
to write to any of the files under /sys/class/pwm/, so you need to use sudo.  

I just added some error checking and pushed it to my repo.  Should I open a 
pull request?

> 
> 
> At Mon, 04 Mar 2019 11:33:24 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:
> 
> > 
> > 
> > balazsracz commented on this pull request.
> > 
> > 
> > 
> > > +        strcat(enablename,"/enable");
> > +        return get_sysfs_file_value(enablename) == 1;
> > +    }
> > +private:
> > +    const uint32_t chip_;
> > +    const uint32_t channel_;
> > +    char pwmdir[40];
> > +    uint32_t get_sysfs_file_value(const char* basename) {
> > +        uint32_t value;
> > +        FILE *fp = fopen(basename,"r");
> > +        fscanf(fp,"%d",&value);
> > +        fclose(fp);
> > +        return value;
> > +    }
> > +    void set_sysfs_file_value(const char* basename, uint32_t value) {
> > +        FILE *fp = fopen(basename,"w");
> > 
> > see https://github.com/RobertPHeller/openmrn/pull/4
> > note you also need to merge openmrn master for it to work.
> > 
> 

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services

RobertPHeller commented:
At Mon, 04 Mar 2019 13:54:01 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:

> 
> 
> > Also, it appears my repo is 30 commits behind bakerstu/openmrn and I have no clue how to fix that. GitHub will only let me make a pull request and is not giving me a way to get the updates. I am about to completely trash my repo and make a fresh fork and start over.
> 
> You have a couple options:
> 1) https://github.com/KirstieJane/STEMMRoleModels/wiki/Syncing-your-fork-to-the-original-repository-via-the-browser
> 2) https://help.github.com/en/articles/syncing-a-fork
> 
> I use the second option normally as I use commandline client often.

Thanks.  This should help...

> 

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services

atanisoft commented:
> One other thing: the way the current sysfs interface work, you have to be root
> to write to any of the files under /sys/class/pwm/, so you need to use sudo

One other option is the setuid bit on the exe. Something like chmod 1755 exec.

atanisoft commented:
@RobertPHeller it looks like they have added a fix to the linux kernel for RPI to allow access when the user has the group "gpio" [ref](https://www.raspberrypi.org/forums/viewtopic.php?t=194174). I don't know if this will work or not but it might be worth exploring as an another alternative to root/sudo.

RobertPHeller commented:
At Tue, 05 Mar 2019 05:58:09 -0800 bakerstu/openmrn <reply@reply.github.com> wrote:

> 
> 
> @RobertPHeller it looks like they have added a fix to the linux kernel for 
> RPI to allow access when the user has the group "gpio" 
> [ref](https://www.raspberrypi.org/forums/viewtopic.php?t=194174). I don't 
> know if this will work or not but it might be worth exploring as an another 
> alternative to root/sudo.

OK, it looks like I need to get a newer kernel for my Pis...  Time to truck 
them to the library for software updates...  (My home Internet is very slow.)

> 

-- 
Robert Heller             -- 978-544-6933
Deepwoods Software        -- Custom Software Services
http://www.deepsoft.com/  -- Linux Administration Services
heller@deepsoft.com       -- Webhosting Services
